spec/services/get_products_service_spec.rb

```diff
@@ -29,146 +33,239 @@
-                                        thumbnail: milk_tea.thumbnail,
-                                        tag: {
-                                          color: nil,
-                                          name: nil
-                                        }
-                                      }])
+                products: [expected_milk_tea]
+              },
+              {
+                name: 'Hi Tea category',
+                products: [expected_hi_tea]
               }
             ]
           end
-          it 'returns children categories having products' do
+
+          it 'returns categories having products' do
             service.call
 
             expect(service.success?).to eq true
-            expect(service.result.size).to eq 2
-            expect(service.result).to match_array(expected_result)
+            expect(service.result.size).to eq 3
+            expect(service.result.to_json).to eq(expected_result.to_json)
+          end
+        end
+
+        context 'when a category has a parent category' do
+          context 'when a category has children categories' do
+            let(:category_id) { tea_category.id }
+            let(:expected_result) do
+              [
+                {
+                  name: 'Hi Tea category',
+                  products: [expected_hi_tea]
+                },
+                {
+                  name: 'Milk Tea category',
+                  products: [expected_milk_tea]
+                }
+              ]
+            end
+
+            it 'returns children categories having products' do
+              service.call
+
+              expect(service.success?).to eq true
+              expect(service.result.size).to eq 2
+              expect(service.result.to_json).to eq(expected_result.to_json)
+            end
+          end
+
+          context 'when a category does not have children categories' do
+            let(:category_id) { milk_tea_category.id }
+            let(:expected_result) do
+              [
+                {
+                  name: 'Milk Tea category',
+                  products: [expected_milk_tea]
+                }
+              ]
+            end
+
+            it 'returns that category having products' do
+              service.call
+
+              expect(service.success?).to eq true
+              expect(service.result.size).to eq 1
+              expect(service.result.to_json).to eq(expected_result.to_json)
+            end
+          end
+        end
+      end
+
+      context 'when products within category does not exist in cache' do
+        include_context 'redis mock'
+
+        let(:cached_value) do
+          JSON.parse(redis.get(category_id)).map do |category|
+            {
+              name: category['name'],
+              products: category['products'].map do |product|
+                {
+                  id: product['id'],
+                  name: product['name'],
+                  price: product['price'].to_f,
```

Comment:
The reason why I have to convert the price as the string to the float

![image](https://github.com/user-attachments/assets/17597ff5-bc4d-4ada-b2c2-4a35550f3e48)